llvm/lib/Transforms/Scalar/EarlyCSE.cpp

```diff
@@ -1632,6 +1632,9 @@ bool EarlyCSE::processNode(DomTreeNode *Node) {
           LLVM_DEBUG(dbgs() << "Skipping due to debug counter\n");
           continue;
         }
+
+        // Potential TODO: We may be throwing away attribute information when
+        // we delete Inst that we could propagate too InVal.first.
```

Comment:
I suspect we will run into correctness issues if we don't combine the attributes after CSE. But I cannot think of a counterexample right now.

```suggestion
        // we delete Inst that we could propagate to InVal.first.
```

Reply:
If we dont combine?

I was thinking about it, are there any **callsite** attributes that can create side-effects?
AFAICT no.

Reply:
We still need to intersect attributes here. If say the first call returns `nonnull` but is (dynamically) unused, and the second one does not return `nonnull` but is used, you'll now replace it with a `nonnull` call and a potential poison value.

Reply:
Okay, I can see that argument for return values, but for param or fn?

Reply:
If we did CSE for non-readonly funtions ofc I could see param args mattering, but with the limits I think its only return?